### ISSUE TITLE
fairmq-shmonit continues to run after the device is shut down

### ISSUE THREAD

**Issue by AndreyLebedev** [FairRootGroup/FairMQ#331]
**Describe the bug**
A `fairmq-shmonit` continues to run after the device and topology are shut down.
A bug was reproduced on the EPN cluster in Milstoneweek 2 tests.
The bug is reproduced reliably.
The DDS topology consists of `TfBuilder` devices only. One `TfBuilder` per node.

**Logs / Screenshots**
<img width="623" alt="Screenshot 2021-02-18 at 17 30 45" src="https://user-images.githubusercontent.com/6470686/108389920-6e452300-7210-11eb-8dc8-b3c11e15eb35.png">

**System information (please complete the following information):**
 - OS: Centos 8

**Comment by rbx** [comment]
Is `fairmq-shmmonitor` run manually, or auto-launched by the device(s)?

For how long after the shutdown does `fairmq-shmmonitor` remain?

A debug log file of `TfBuilder` would help.

**Comment by AndreyLebedev** [comment]
1) It's auto-launched by the device.
2) It remains for at least a couple of minutes and then I kill it manually. Probably it will stay longer if not killed.
3) This is the end of the log. The full log is big and I guess unrelated
```
[2021-02-18 17:20:11.843][E] BUG: RegionAllocatorResource: REPEATED INSERT!!!  0x7f4e401eaeb0 : 96, original size: 192
[2021-02-18 17:20:11.910][I] TfBuilderInput::stop: Merger queue emptied.
[2021-02-18 17:20:11.910][I] TfBuilderInput::stop: Merger thread stopped.
[2021-02-18 17:20:11.910][I] TfBuilderInput: Teardown complete.
[2021-02-18 17:20:11.910][D] <tfb-main> TfBuilderDevice::stop(): Input handler stopped.
[2021-02-18 17:20:11.910][D] <tfb-main> TfBuilderDevice::stop(): Forward thread stopped.
[2021-02-18 17:20:11.910][D] <tfb-main> TfBuilderDevice::stop(): Info thread stopped.
[2021-02-18 17:20:12.266][D] Exiting TfBuilder Update sending thread.
[2021-02-18 17:20:12.266][D] Exiting Stf requesting thread.
[2021-02-18 17:20:12.275][D] <tfb-main> TfBuilderDevice::stop(): RPC clients stopped.
[2021-02-18 17:20:12.275][D] <tfb-main> TfBuilderDevice() stopped... 
[2021-02-18 17:20:12.275][D] <tfb-main> PostRun()
[2021-02-18 17:20:12.275][I] [STATE][FMQ] RUNNING ---> READY
[2021-02-18 17:20:12.275][D] <tfb-main> [FMQ] Publishing state-change: RUNNING->READY to 15903701552739631229
[2021-02-18 17:20:22.547][I] [STATE][FMQ] READY ---> RESETTING TASK
[2021-02-18 17:20:22.547][D] <tfb-main> [FMQ] Publishing state-change: READY->RESETTING TASK to 15903701552739631229
[2021-02-18 17:20:22.547][D] <tfb-main> Sending TfBuilder update. accepting=false total=2011 <msgs_suppressed=2>
[2021-02-18 17:20:22.547][E] TfBuilderUpdate: no gRPC connection to scheduler
[2021-02-18 17:20:22.547][E] Sending TfBuilder status update failed.
[2021-02-18 17:20:22.547][D] <tfb-main> TfBuilderDevice::stop(): mRunning is false.
[2021-02-18 17:20:22.547][E] NumStfSendersInPartitionRequest: no gRPC connection to scheduler
[2021-02-18 17:20:22.547][E] RPC error: Request for StfSender disconnect failed!
[2021-02-18 17:20:22.547][D] <tfb-main> TfBuilderInput::stop: Waiting for input threads to terminate.
[2021-02-18 17:20:22.547][D] <tfb-main> TfBuilderInput::stop: All input threads terminated.
[2021-02-18 17:20:22.547][I] TfBuilderInput::stop: All input channels are closed.
[2021-02-18 17:20:22.547][D] <tfb-main> TfBuilderInput::stop: Stopping the STF merger thread.
[2021-02-18 17:20:22.547][I] TfBuilderInput::stop: Merger queue emptied.
[2021-02-18 17:20:22.547][I] TfBuilderInput::stop: Merger thread stopped.
[2021-02-18 17:20:22.547][I] TfBuilderInput: Teardown complete.
[2021-02-18 17:20:22.547][D] <tfb-main> TfBuilderDevice::stop(): Input handler stopped.
[2021-02-18 17:20:22.547][D] <tfb-main> TfBuilderDevice::stop(): Forward thread stopped.
[2021-02-18 17:20:22.547][D] <tfb-main> TfBuilderDevice::stop(): Info thread stopped.
[2021-02-18 17:20:22.547][D] <tfb-main> TfBuilderDevice::stop(): RPC clients stopped.
[2021-02-18 17:20:22.547][D] <tfb-main> TfBuilderDevice() stopped... 
[2021-02-18 17:20:22.547][D] <tfb-main> ResetTask()
[2021-02-18 17:20:22.547][I] [STATE][FMQ] RESETTING TASK ---> DEVICE READY
[2021-02-18 17:20:22.547][D] <tfb-main> [FMQ] Publishing state-change: RESETTING TASK->DEVICE READY to 15903701552739631229
[2021-02-18 17:20:22.627][I] [STATE][FMQ] DEVICE READY ---> RESETTING DEVICE
[2021-02-18 17:20:22.627][D] <tfb-main> [FMQ] Publishing state-change: DEVICE READY->RESETTING DEVICE to 15903701552739631229
[2021-02-18 17:20:22.627][D] [FMQ] AcksSender for fmq_e2575ba4_rg_1 leaving (blocks left to free: 0,  blocks left to send: 0).
[2021-02-18 17:20:22.820][D] [FMQ] AcksReceiver for fmq_e2575ba4_rg_1 leaving (remaining queue size: 0).
[2021-02-18 17:20:22.820][D] <tfb-main> [FMQ] Region 'fmq_e2575ba4_rg_1' destroyed.
[2021-02-18 17:20:22.820][D] <tfb-main> [FMQ] Region queue 'fmq_e2575ba4_rgq_1' destroyed.
[2021-02-18 17:20:22.820][D] <tfb-main> [FMQ] Region 'fmq_e2575ba4_rg_1' (local) destructed.
[2021-02-18 17:20:22.825][D] [FMQ] AcksSender for fmq_e2575ba4_rg_2 leaving (blocks left to free: 0,  blocks left to send: 0).
[2021-02-18 17:20:22.927][D] [FMQ] AcksReceiver for fmq_e2575ba4_rg_2 leaving (remaining queue size: 0).
[2021-02-18 17:20:22.927][D] <tfb-main> [FMQ] Region 'fmq_e2575ba4_rg_2' destroyed.
[2021-02-18 17:20:22.927][D] <tfb-main> [FMQ] Region queue 'fmq_e2575ba4_rgq_2' destroyed.
[2021-02-18 17:20:22.927][D] <tfb-main> [FMQ] Region 'fmq_e2575ba4_rg_2' (local) destructed.
[2021-02-18 17:20:23.619][D] <tfb-main> [FMQ] Destroying Shared Memory transport...
[2021-02-18 17:20:23.619][D] <tfb-main> [FMQ] Last segment user, removing segment.
Cleaning up for shared memory id 'e2575ba4'...
Region counter found: 2
Did not remove 'fmq_e2575ba4_rg_1'. Already removed?
Did not remove 'fmq_e2575ba4_rgq_1'. Already removed?
Did not remove 'fmq_e2575ba4_rg_2'. Already removed?
Did not remove 'fmq_e2575ba4_rgq_2'. Already removed?
Successfully removed 'fmq_e2575ba4_m_0'.
Successfully removed 'fmq_e2575ba4_mng'.
Successfully removed 'fmq_e2575ba4_mtx'.
Successfully removed 'fmq_e2575ba4_cv'.
[2021-02-18 17:20:23.619][I] [STATE][FMQ] RESETTING DEVICE ---> IDLE
[2021-02-18 17:20:23.620][D] <tfb-main> [FMQ] Publishing state-change: RESETTING DEVICE->IDLE to 15903701552739631229
[2021-02-18 17:21:05.158][I] [FMQ] Received device shutdown request (signal 15).
[2021-02-18 17:21:05.158][I] [FMQ] Waiting for graceful device shutdown. Hit Ctrl-C again to abort immediately.
[2021-02-18 17:21:05.158][I] [STATE][FMQ] IDLE ---> EXITING
[2021-02-18 17:21:05.158][D] <tfb-main> [FMQ] Plugin 'dds' cannot release control because it has no control.
[2021-02-18 17:21:05.158][D] <tfb-main> [FMQ] Publishing state-change: IDLE->EXITING to 15903701552739631229
[2021-02-18 17:21:05.158][E] [FMQ] Uncaught exception reached the top of DeviceRunner: _Map_base::at
[2021-02-18 17:21:05.158][D] <tfb-main> [FMQ] Shutting down Plugin Manager
[2021-02-18 17:21:05.158][D] <tfb-main> [FMQ] Plugin 'dds' cannot release control because it has no control.
[2021-02-18 17:21:06.159][D] <tfb-main> [FMQ] Unloaded plugin: 'dds', version '1.4.30', maintainer 'FairRootGroup <fairroot@gsi.de>', homepage 'https://github.com/FairRootGroup/FairMQ'
[2021-02-18 17:21:06.159][D] <tfb-main> [FMQ] Unloaded plugin: 'control', version '1.4.30', maintainer 'FairRootGroup <fairroot@gsi.de>', homepage 'https://github.com/FairRootGroup/FairMQ'
[2021-02-18 17:21:06.159][D] <tfb-main> [FMQ] Unloaded plugin: 'config', version '1.4.30', maintainer 'FairRootGroup <fairroot@gsi.de>', homepage 'https://github.com/FairRootGroup/FairRoot'
[2021-02-18 17:21:06.159][D] <tfb-main> [FMQ] Shutting down Plugin Services
[2021-02-18 17:21:06.159][D] <tfb-main> TfBuilderDevice::~TfBuilderDevice()
[2021-02-18 17:21:06.159][D] <tfb-main> (Sub)TimeFrame Source terminated...
[2021-02-18 17:21:06.159][D] <tfb-main> (Sub)TimeFrame Sink terminated.
[2021-02-18 17:21:06.159][D] <tfb-main> Erasing DataDistribution discovery key: epn/data-dist/partition/2PMrW3foY4i/TfBuilder/TfBuilder-epn025-278a5899
[2021-02-18 17:21:06.160][D] <tfb-main> [FMQ] Shutting down device tfb
[2021-02-18 17:21:06.160][I] [STATE][FMQ] Exiting FairMQ state machine
```

**Comment by rbx** [comment]
Thanks, could you also post the beginning of the log? The initialization part, i guess until the repetitive unrelated stuff starts.

**Comment by rbx** [comment]
I think I can reproduce it. Seems to be a regression in the latest tag. Investigating.

**Comment by AndreyLebedev** [comment]
> Thanks, could you also post the beginning of the log? The initialization part, i guess until the repetitive unrelated stuff starts.

Do you still need logs?

**Comment by rbx** [comment]
@AndreyLebedev No, thanks! I found the problem. Will post a fix in a bit.

**Comment by rbx** [comment]
@AndreyLebedev This is now fixed in v1.4.31.
